### PR TITLE
[jit] support tensor as key type in script

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23640 [WIP] add a jit version of adagrad
* #23842 [jit] support grad and data attribute for tensor in script
* #23841 [jit] add initial support for sparse tensors
* #23639 [jit] support dict augment assign in script
* **#23638 [jit] support tensor as key type in script**

Summary:
We add support for tensor as key type in both torch::Dict and
TorchScript, the hash comparision matches the python behavior
(no tp_hash specified when the torch.Tensor bind with python)
where we take the hash of the TensorImpl address as unique identifier.

Test Plan:
jit tests and dict_test.cpp

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23638

Differential Revision: [D16683557](https://our.internmc.facebook.com/intern/diff/D16683557)

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
We have to be careful with making Tensor a key type because constant pooling / common subexpression elimination / storing of Tensors when saving & loading may make two tensors which were a different pointer in python the same pointer in torchscript.

**Comment by wanchaol** [comment]
> We have to be careful with making Tensor a key type because constant pooling / common subexpression elimination / storing of Tensors when saving & loading may make two tensors which were a different pointer in python the same pointer in torchscript.

shouldn't alias DB treat the dictionary as mutable or potentiallly have writers, then constant pooling and cse just ignore them?

**Comment by eellison** [comment]
> > We have to be careful with making Tensor a key type because constant pooling / common subexpression elimination / storing of Tensors when saving & loading may make two tensors which were a different pointer in python the same pointer in torchscript.
> 
> shouldn't alias DB treat the dictionary as mutable or potentiallly have writers, then constant pooling and cse just ignore them?

What I did for https://github.com/pytorch/pytorch/pull/19576 should be more or less the same correctness conditions. But since then a lot of changes have been made so I think it should be relooked at. Serialization, Classtypes, potentially checking for uses in a user defined function, idk

**Comment by eellison** [comment]
Also the assumptions may need to be revisited about inlining may need to be revisited / future proofed

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -842,6 +842,7 @@ struct CAFFE2_API DictType : public Type {
       case TypeKind::IntType:
       case TypeKind::FloatType:
       case TypeKind::StringType:
+      case TypeKind::TensorType:
         return DictTypePtr(new DictType(key, value));
       default:
         AT_ERROR(
```
Comment (zdevito) [inline aten/src/ATen/core/jit_type.h]: only int, float, Tensor and string keys...

**Review by zdevito — APPROVED** [review]
I think we are ok with regard to constant deduplication. Putting a tensor into a map as a key should make it the tensor alias the heap, meaning that we must infer it could be written to. If we think a write to a tensor is possible then it can't be a constant subject to deduplication anyway. It is possible that somewhere in constant handling we have implemented constant de-dup incorrectly, but that is bug in that code and not in this code.

**Comment by kostmo** [comment]
Looks like this broke `caffe2_py2_android_ubuntu16_04_build` in `master`.  Is there a quick fix?

**Comment by wanchaol** [comment]
@kostmo I am testing if https://github.com/pytorch/pytorch/pull/24071 works

**Pull request closed without merging**